### PR TITLE
Schedule tasks of cleared dags

### PR DESCRIPTION
**Pull request by soltanianalytics**
See https://github.com/apache/airflow/issues/13407

This is alternative 2: Keep the `max_active_runs` restriction*, but run `DagRun`s in order of `execution_date` instead of using `TI` to infer what `running` `DagRun`s are really _active_ and which are merely _running_.

*This could be violated in exactly one way:
- Have active `n` `DagRun`s with running task(s)
- Clear more than `max_active_runs - n` `DagRun`s that have `execution_date` before the previously running `DagRun`s
- Then the scheduler will schedule tasks from the first `max_active_runs` tasks, but there will be already running tasks from other `DagRun`s that were previously already running (no new tasks will be scheduled, but the existing ones will continue running)

### CODE REVIEW

**Comment by soltanianalytics** [comment]
This and https://github.com/apache/airflow/pull/13408 are mutually exclusive

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -1478,42 +1478,49 @@ def _do_scheduling(self, session) -> int:
                 guard.commit()
                 # END: create dagruns
 
-            dag_runs = DagRun.next_dagruns_to_examine(session)
+            dag_runs = list(DagRun.next_dagruns_to_examine(session))
```
Comment (turbaszek) [inline airflow/jobs/scheduler_job.py]: Is this necessary?

Comment (soltanianalytics) [reply airflow/jobs/scheduler_job.py]: Indeed, it is not. Removing it

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -1483,37 +1483,46 @@ def _do_scheduling(self, session) -> int:
...
+            # This query will violate max_active_runs by exactly one if tasks in
+            # max_active_runs or more DAGs are cleared while another DAG is
+            # running. It is to deal with the case when a user
             # clears more than max_active_runs older tasks -- we don't want the
             # scheduler to suddenly go and start running tasks from all of the
             # runs. (AIRFLOW-137/GH #1442)
-            #
-            # The longer term fix would be to have `clear` do this, and put DagRuns
-            # in to the queued state, then take DRs out of queued before creating
-            # any new ones
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: Please keep this comment

**Review by ashb — CHANGES_REQUESTED** [review]
This needs unit tests adding to prevent future regressions

File: airflow/jobs/scheduler_job.py (hunk 2 above)
Comment (soltanianalytics) [reply airflow/jobs/scheduler_job.py]: done

**Comment by soltanianalytics** [comment]
> This needs unit tests adding to prevent future regressions

I tried writing a test for this but didn't get too far. Could you hint me at how and what do to?

I thought about something along those lines, but I'm pretty sure I'm missing some core concepts and in any case, it's not finished:

```Python
    def test_do_schedule_cleared_tasks_with_max_active_runs(self):
        """
        Test that the scheduler schedules tasks in DagRun in the order of execution
        date if tasks in more than max_active_runs DagRuns were cleared.
        """
        schedule_interval = timedelta(hours=1)
        task_id = "dummy1"

        job = SchedulerJob(subdir=os.devnull)
        job.executor = MockExecutor(do_update=False)
        job.processor_agent = mock.MagicMock(spec=DagFileProcessorAgent)

        with DAG(
            dag_id='test_schedule_cleared_tasks_with_max_active_runs',
            start_date=DEFAULT_DATE,
            end_date=DEFAULT_DATE + 3 * schedule_interval,
            schedule_interval=schedule_interval,
            max_active_runs=1,
        ) as dag:
            # Cant use DummyOperator as that goes straight to success
            task1 = BashOperator(task_id=task_id, bash_command='true')

        session = settings.Session()
        dagbag = DagBag(
            dag_folder=os.devnull,
            include_examples=False,
            read_dags_from_db=True,
        )
        dagbag.bag_dag(dag=dag, root_dag=dag)
        dagbag.sync_to_db(session=session)

        # Create three DagRuns with different execution_dates
        dag_runs = [
            dag.create_dagrun(
                run_type=DagRunType.SCHEDULED,
                execution_date=DEFAULT_DATE + i * schedule_interval,
                state=State.RUNNING,
                session=session,
            ) for i in range(3)
        ]
        dagbag.sync_to_db(session=session)

        # queue the task for the first DagRun, and no other
        assert job._do_scheduling(session) == 1
        assert dag_runs[0].get_task_instance(task_id).state == State.QUEUED
        assert dag_runs[1].get_task_instance(task_id).state == State.NONE
        assert dag_runs[2].get_task_instance(task_id).state == State.NONE

        # Do something so that the first two DagRuns succeed and the third is running
        # ???


        # Clear the first two DagRuns & their tasks somehow
        # ???

        # Test that job._do_scheduling(session) only queue the first TI
        assert job._do_scheduling(session) == 1
        assert dag_runs[0].get_task_instance(task_id).state == State.QUEUED
        assert dag_runs[1].get_task_instance(task_id).state == State.NONE
        assert dag_runs[2].get_task_instance(task_id).state == State.RUNNING
```

**Comment by kaxil** [comment]
The following tests are failing:

```
FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_0
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_1_up_for_retry
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_2_up_for_reschedule
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_depends_on_past_0
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_depends_on_past_1_up_for_retry
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_depends_on_past_2_up_for_reschedule
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_with_task_concurrency_0
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_with_task_concurrency_1_up_for_retry
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_dag_file_processor_process_task_instances_with_task_concurrency_2_up_for_reschedule
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_runs_respected_after_clear
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_scheduler_job_add_new_task
  FAILED tests/jobs/test_scheduler_job.py::TestDagFileProcessor::test_should_mark_dummy_task_as_success
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_dagrun_callbacks_are_called_0_success
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_dagrun_callbacks_are_called_1_failed
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_dagrun_callbacks_are_not_added_when_callbacks_are_not_defined_0_success
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_dagrun_callbacks_are_not_added_when_callbacks_are_not_defined_1_failed
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_do_schedule_max_active_runs_upstream_failed
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_scheduler_verify_pool_full
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_scheduler_verify_pool_full_2_slots_per_task
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_scheduler_verify_priority_and_slots
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_verify_integrity_if_dag_changed
  FAILED tests/jobs/test_scheduler_job.py::TestSchedulerJob::test_verify_integrity_if_dag_not_changed
  = 22 failed, 2065 passed, 97 skipped, 1 xfailed, 1 warning in 280.68s (0:04:40) =
```

**Comment by jhtimmins** [comment]
@soltanianalytics do you need any help moving this along or have questions about what needs to get done?

**Comment by soltanianalytics** [comment]
Frankly, I realized that I understood the airflow code base much less than I thought and I didn't continue with this. I won't be able to contribute in the near future so I'll close it for now. Thanks for reaching out!

**Comment by jhtimmins** [comment]
Thanks for the update @soltanianalytics! If you want to take another crack at contributing in the future, I'm happy to help you find issues/bugs that are more approachable.

**Pull request closed without merging**